File: searchIn2DMatrixII.java

// Time Complexity : o(m + n)
// Space Complexity : o(1)
// Did this code successfully run on Leetcode : yess
// Three line explanation of solution in plain english
// simple problem actually; can be done on observation; we can start on m-1,0 index or 0,n-1 index 
// we know deterministically that elemets in particular direction behave in a certain way
// Your code here along with comments explaining your approach

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix == null || matrix.length == 0) {
            return false;
        }
        int m = matrix.length - 1;
        int n = matrix[0].length - 1;
        int i = m - 1;
        int j = 0;
        while(i >= 0 && j < n) {
            if(matrix[i][j] == target) {
                return true;
            }
            if(matrix[i][j] > target) {
                i--;
            }
            else {
                j++;
            }
        }
        return false;
    }
}

File: SearchIn2DMatrixII.py
# Time Complexity: O(rows+columns)
# Space Complexity: O(1)
# Successfully ran on leetcode
 
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows = len(matrix)
        cols = len(matrix[0])
        i = 0
        j = cols-1
        while i<rows and j>=0:
            if matrix[i][j]==target:
                return True
            elif target>matrix[i][j]:
                i+=1
            elif target<matrix[i][j]:
                j-=1
        return False